### ISSUE TITLE
Duplicate conflicting deployments performed when enabling DeployAllMultipleTemplateParameterFiles and ParallelDeployMultipleTemplateParameterFiles

### ISSUE THREAD

**Issue by Xitric** [Azure/AzOps#886]
**Describe the bug**

In our AzOps repository, we have enabled both `AllowMultipleTemplateParameterFiles`, `DeployAllMultipleTemplateParameterFiles` and `ParallelDeployMultipleTemplateParameterFiles`. This works great the majority of the time, but sometimes it causes deployments to fail.

Firstly, we noticed that the validation pipeline would report the What-If results on our deployments twice:

```
WhatIf Results for main.json with main.instance01.parameters.json:
...

WhatIf Results for main.json with main.instance01.parameters.json:
...
```

In the pipeline output we can also see that it is processing a separate deployment for both the template and the parameter file:

```
[[14:20:15][New-AzOpsDeployment] Processing deployment AzOps-main.instance01-B9EE for template /__w/1/s/root/mg (e3e8e9e9-ee3c-4b13-abd2-12efc9f08498)/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/main.json with parameter "/__w/1/s/root/mg (e3e8e9e9-ee3c-4b13-abd2-12efc9f08498)/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/main.instance01.parameters.json" in mode Incremental
14:20:15][New-AzOpsDeployment] Processing deployment AzOps-main.instance01.parameters-B9EE for template /__w/1/s/root/mg (e3e8e9e9-ee3c-4b13-abd2-12efc9f08498)/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/main.json with parameter "/__w/1/s/root/mg (e3e8e9e9-ee3c-4b13-abd2-12efc9f08498)/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/main.instance01.parameters.json" in mode Incremental
Getting the latest status of all resources...Getting the latest status of all resources...
```

Sometimes this succeeds, and other times it fails with the error:

> Get-AzDeploymentWhatIfResult: The process cannot access the file '/__w/1/s/root/mg (e3e8e9e9-ee3c-4b13-abd2-12efc9f08498)/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/main.instance01.parameters.json' because it is being used by another process.

In the case described above, only the Bicep template was modified and not the corresponding parameter file.

**Importantly**, it should be noted that so far, we have only observed this behaviour when we have a 1-1 relationship between Bicep templates and parameter files. We perform a large number of deployments where we have numerous parameter files per Bicep template, and in these cases we haven't observed duplicate deployments.

TL;DR: If we change a Bicep template that is associated with multiple parameter files, deployments are only processed for the parameter files and not the Bicep template itself. If we change a Bicep template that is associated with a single parameter file, deployments are only processed for both the parameter file and the Bicep template itself.

**Steps to reproduce**

1. Enable `AllowMultipleTemplateParameterFiles`, `DeployAllMultipleTemplateParameterFiles` and `ParallelDeployMultipleTemplateParameterFiles`
2. Ensure that `ThrottleLimit` is at least 2
3. Create a Bicep template along with a single associated parameter file
4. Create a PR that modifies the Bicep template, and observe how the validation pipeline processes the same deployment twice, in parallel

**Comment by Xitric** [comment]
@Jefajers I will try to create a minimally reproducible example from the accelerator repo using our configurations to hopefully let you reproduce this consistently. Perhaps we have yet another configuration that is causing this, which I have overlooked.

I can assure you that we do not have any lingering parameter conversion files in git.

**Comment by Jefajers** [comment]
Thanks it would be very valuable to understand the given situation I appreciate your assistance🤓.

**Comment by Xitric** [comment]
@Jefajers I got a lot closer to what is causing this issue for us. Here is what I did:

1. Enable
    - `Core.AllowMultipleTemplateParameterFiles`
    - `Core.DeployAllMultipleTemplateParameterFiles`
    - `Core.ParallelDeployMultipleTemplateParameterFiles`
2. Ensure `ThrottleLimit` is at least 2
3. Create the following files on `main`
    - `a.westeurope.bicep`
    - `a.westeurope.x1.bicepparam`
    - `b.westeurope.bicep`
4. Create a PR that modifies `a.westeurope.bicep` _and_ `b.westeurope.bicep`
5. Observe that AzOps performs two deployments of `a.westeurope.json` in parallel:
    ```
    [Invoke-AzOpsPush] Deployment required
    [Invoke-AzOpsPush] Adding or modifying:
    [Invoke-AzOpsPush]   root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.bicep
    [Invoke-AzOpsPush]   root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/b.westeurope.bicep
    [New-AzOpsDeployment] Processing deployment AzOps-a.westeurope.x1.parameters-B9EE for template /home/vsts/work/1/s/root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.json with parameter "/home/vsts/work/1/s/root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.x1.parameters.json" in mode Incremental
    Getting the latest status of all resources...[13:39:58][New-AzOpsDeployment] Processing deployment AzOps-a.westeurope.x1-B9EE for template /home/vsts/work/1/s/root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.json with parameter "/home/vsts/work/1/s/root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.x1.parameters.json" in mode Incremental
    Getting the latest status of all resources...
    
    
    
    New-AzResourceGroupDeployment: The process cannot access the file '/home/vsts/work/1/s/root/sub (839f6ab6-4e09-4489-a2ff-b5ce962a600c)/my-resource-group/a.westeurope.x1.parameters.json' because it is being used by another process.
    ```

If I disable `Core.DeployAllMultipleTemplateParameterFiles`, the issue ceases to occur.

If I rename the files to something like this, the issue also ceases to occur and only one deployment is performed:
- `a.bicep`
- `a.x1.bicepparam`
- `b.bicep`

So it has something to do with how we name our templates and parameter files, and how AzOps interprets that.

Creating a file `b.westeurope.x1.bicepparam` does not solve the issue, so it is not caused by having a lingering template with no parameter file.

**Comment by Jefajers** [comment]
Hi again @Xitric, thanks given the above scenario I am now able to reproduce the result without any unexpected manipulation :relieved:.

I have an ide of how to combine the previous suggestion of streamlining the deployment name together with further validations to not only get the expected amount of deployments but even more importantly I think we might have an additional issue that could result in template pairs not even changed in the PR could be queued up for deployment which is never expected to happen.

Will update this issue with associate PR once I have a more complete suggestion of a fix :ambulance:.
